pkg/endpoint/endpoint.go

```diff
@@ -1202,40 +1224,138 @@ func (e *Endpoint) GetDockerNetworkID() string {
+		_, fileName, fileLine, _ := runtime.Caller(1)
+		log.Infof("State transition skipped: %s -> %s at %s:%d", e.state, toState, fileName, fileLine)
+	}
+	return false
+
+OKState:
+	e.state = toState
+	return true
 }
 
-// bumpPolicyRevision marks the endpoint to be running the next scheduled
-// policy revision as setup by e.regenerate(). endpoint.Mutex may not be held
-func (e *Endpoint) bumpPolicyRevision() {
-	e.Mutex.Lock()
-	e.policyRevision = e.nextPolicyRevision
-	e.updateLogger()
-	e.Mutex.Unlock()
+// BuilderSetStateLocked modifies the endpoint's state
+// endpoint.Mutex must be held
+// endpoint BuildMutex must be held!
+func (e *Endpoint) BuilderSetStateLocked(toState string) bool {
+	// Validate the state transition.
+
+	switch e.state { // From state
+	case StateCreating, StateWaitingForIdentity, StateReady, StateDisconnecting, StateDisconnected:
+		// No valid transitions for the builder
+	case StateWaitingToRegenerate:
+		switch toState {
+		// Builder transitions the endpoint from
+		// waiting-to-regenerate state to regenerating state
+		// right after aquiring the endpoint lock, and while
+		// endpoint's build mutex is held. All changes to
+		// cilium and endpoint configuration, policy as well
+		// as the existing set of security identities will be
+		// reconsidered after this point, i.e., even if some
+		// of them are changed regeneration need not be queued
+		// if the endpoint is already in waiting-to-regenerate
+		// state.
+		case StateRegenerating:
+			goto OKState
+		}
+	case StateRegenerating:
+		switch toState {
+		// While still holding the build mutex, the builder
+		// tries to transition the endpoint to ready
+		// state. But since the endpoint mutex was released
+		// for the duration of the bpf generation, it is
+		// possible that another build request has been
+		// queued. In this case the endpoint has been
+		// transitioned to waiting-to-regenerate state
+		// already, and the transition to ready state is
+		// skipped.
+		case StateReady:
+			goto OKState
+		}
+	}
+	return false
+
+OKState:
+	e.state = toState
+	return true
 }
 
-// IsDisconnecting returns true if the endpoint is being disconnected or
-// already disconnected
-// endpoint.Mutex may only be RLock()ed
-func (e *Endpoint) IsDisconnecting() bool {
-	e.Mutex.RLock()
-	defer e.Mutex.RUnlock()
-	return e.IsDisconnectingLocked()
+// SetStateForTestingOnly modifies the endpoint's state
+// This is used when creating endpoints for testing, so no locking is done.
+func (e *Endpoint) SetStateForTestingOnly(toState string) {
+	e.state = toState
 }
 
-// IsDisconnectingLocked is identical to IsDisconnecting but with the endpoint
-// lock held for reading or writing
-func (e *Endpoint) IsDisconnectingLocked() bool {
-	return e.State == StateDisconnected || e.State == StateDisconnecting
+// bumpPolicyRevision marks the endpoint to be running the next scheduled
+// policy revision as setup by e.regenerate(). endpoint.Mutex may not be held
```

Comment:
s/  may not be held/ should not be held/

Reply:
Changed, but really it "must not be held", as it will deadlock otherwise :-)